File: frontend/src/services/offerService.js
import { api } from './api';

/**
 * Fetches list of categories
 * @returns {Promise<Array>} List of categories
 */
export const fetchCategories = async () => {
  try {
    // First try to get categories from the API
    const response = await api.get('/categories');
    return response.data;
  } catch (error) {
    console.error('Error fetching categories:', error);
    
    // Fallback to mock data if the API fails
    if (process.env.NODE_ENV === 'development') {
      console.warn('Falling back to mock data for categories');
      return {
        items: [
          { id: 1, name: "Electronics" },
          { id: 2, name: "Books" },
          { id: 3, name: "Clothing" },
          { id: 4, name: "Home" },
          { id: 5, name: "Games" }
        ]
      };
    }
    
    throw new Error('Nie udało się pobrać kategorii');
  }
};

/**
 * Fetches paginated list of offers with optional filters
 * @param {Object} params - Query parameters
 * @param {string} [params.search] - Search term for offer title/description
 * @param {number} [params.category_id] - Filter by category ID
 * @param {number} [params.page=1] - Page number
 * @param {number} [params.limit=20] - Number of items per page
 * @param {string} [params.sort='created_at_desc'] - Sort order
 * @returns {Promise<Object>} Paginated list of offers
 */
export const fetchOffers = async (params = {}) => {
  try {
    // Use POST request to /offers/search endpoint which we implemented in the backend
    const response = await api.post('/offers/search', {
      search: params.search,
      category_id: params.category_id ? parseInt(params.category_id) : undefined,
      page: params.page || 1,
      limit: params.limit || 20,
      sort: params.sort || 'created_at_desc'
    });
    return response.data;
  } catch (error) {
    console.error('Error fetching offers:', error);
    
    // Fallback to mock data if the API fails
    if (process.env.NODE_ENV === 'development') {
      console.warn('Falling back to mock data for offers');
      return getMockOffers(params);
    }
    
    throw new Error('Nie udało się pobrać ofert');
  }
};

/**
 * Helper function to provide mock data when the API is unavailable
 * @private
 */
const getMockOffers = (params = {}) => {
  const mockOffers = [
    {
      id: "550e8400-e29b-41d4-a716-446655440000",
      title: "PlayStation 5",
      description: "Brand new PlayStation 5 console with controller",
      price: "2499.99",
      image_filename: null,
      category_id: 1,
      quantity: 5,
      status: "active",
      created_at: "2025-05-10T12:00:00Z",
      seller_id: "550e8400-e29b-41d4-a716-446655440001"
    },
    {
      id: "550e8400-e29b-41d4-a716-446655440002",
      title: "The Lord of the Rings",
      description: "Complete trilogy book set",
      price: "199.99",
      image_filename: null,
      category_id: 2,
      quantity: 10,
      status: "active",
      created_at: "2025-05-09T10:00:00Z",
      seller_id: "550e8400-e29b-41d4-a716-446655440001"
    },
    {
      id: "550e8400-e29b-41d4-a716-446655440003",
      title: "Nike Running Shoes",
      description: "Premium running shoes size 42",
      price: "299.99",
      image_filename: null,
      category_id: 3,
      quantity: 3,
      status: "active",
      created_at: "2025-05-08T14:30:00Z",
      seller_id: "550e8400-e29b-41d4-a716-446655440004"
    }
  ];

  // Filter by search term if provided
  let filteredOffers = [...mockOffers];
  if (params.search) {
    const searchTerm = params.search.toLowerCase();
    filteredOffers = filteredOffers.filter(
      offer => 
        offer.title.toLowerCase().includes(searchTerm) || 
        (offer.description && offer.description.toLowerCase().includes(searchTerm))
    );
  }

  // Filter by category if provided
  if (params.category_id) {
    filteredOffers = filteredOffers.filter(
      offer => offer.category_id === parseInt(params.category_id)
    );
  }

  // Sort offers
  if (params.sort === 'price_asc') {
    filteredOffers.sort((a, b) => parseFloat(a.price) - parseFloat(b.price));
  } else if (params.sort === 'price_desc') {
    filteredOffers.sort((a, b) => parseFloat(b.price) - parseFloat(a.price));
  } else {
    // Default to created_at_desc
    filteredOffers.sort((a, b) => new Date(b.created_at) - new Date(a.created_at));
  }

  // Paginate
  const page = params.page || 1;
  const limit = params.limit || 20;
  const startIndex = (page - 1) * limit;
  const endIndex = page * limit;
  const paginatedOffers = filteredOffers.slice(startIndex, endIndex);

  return {
    items: paginatedOffers,
    total: filteredOffers.length,
    page: page,
    limit: limit,
    pages: Math.ceil(filteredOffers.length / limit)
  };
};

/**
 * Creates a new offer
 * @param {FormData} formData - Form data containing offer details and image
 * @returns {Promise<Object>} Created offer data
 */
export const createOffer = async (formData) => {
  try {
    const response = await api.post('/offers', formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    });
    return response.data;
  } catch (error) {
    console.error('Error creating offer:', error);
    throw new Error(
      error.response?.data?.message || 'Nie udało się utworzyć oferty'
    );
  }
};

/**
 * Fetches list of seller's offers
 * @param {Object} params - Query parameters
 * @param {number} params.page - Page number
 * @param {number} params.limit - Number of items per page
 * @param {string} params.sort - Sort order
 * @returns {Promise<Object>} Paginated list of offers
 */
export const fetchSellerOffers = async (params = {}) => {
  try {
    const response = await api.get('/seller/offers', { params });
    return response.data;
  } catch (error) {
    console.error('Error fetching seller offers:', error);
    throw new Error('Nie udało się pobrać ofert');
  }
};

/**
 * Fetches details of a single offer
 * @param {string} offerId - ID of the offer to fetch
 * @returns {Promise<Object>} Offer details
 */
export const fetchOfferDetails = async (offerId) => {
  try {
    // First try to get the offer details using the search API
    try {
      console.log(`Fetching offer details for ID: ${offerId}`);
      
      // Use the offers/search endpoint with pagination
      const response = await api.post('/offers/search', {
        page: 1,
        limit: 20
      });
      
      console.log('Search response:', response.data);
      
      // If we got results, find the specific offer by ID
      if (response.data && response.data.items && response.data.items.length > 0) {
        const offer = response.data.items.find(item => item.id === offerId);
        
        if (offer) {
          console.log('Found offer in search results:', offer);
          
          // Fetch categories to get category name
          const categoriesResponse = await fetchCategories();
          const categories = categoriesResponse.items || [];
          
          const category = categories.find(cat => cat.id === offer.category_id) || {
            id: offer.category_id,
            name: getCategoryNameById(offer.category_id)
          };
          
          // Enrich the offer data with seller and category information
          return {
            ...offer,
            seller: {
              id: offer.seller_id,
              first_name: "Seller",
              last_name: `#${offer.seller_id.substring(0, 6)}`
            },
            category: category
          };
        } else {
          console.warn(`Offer with ID ${offerId} not found in search results`);
        }
      }
      
      // If no results or offer not found, continue to fallback
      console.warn('Using fallback for offer details');
    } catch (apiError) {
      console.warn('API error when fetching offer details:', apiError);
    }
    
    // Fallback to mock data
    console.log('Using mock data for offer details');
    const mockOffers = {
      "550e8400-e29b-41d4-a716-446655440000": {
        id: "550e8400-e29b-41d4-a716-446655440000",
        title: "PlayStation 5",
        description: "Brand new PlayStation 5 console with controller. Includes two games and an extra controller. Perfect condition, never used.",
        price: "2499.99",
        image_filename: null,
        category_id: 1,
        quantity: 5,
        status: "active",
        created_at: "2025-05-10T12:00:00Z",
        updated_at: null,
        seller_id: "550e8400-e29b-41d4-a716-446655440001",
        seller: {
          id: "550e8400-e29b-41d4-a716-446655440001",
          first_name: "John",
          last_name: "Seller"
        },
        category: {
          id: 1,
          name: "Electronics"
        }
      },
      "550e8400-e29b-41d4-a716-446655440002": {
        id: "550e8400-e29b-41d4-a716-446655440002",
        title: "The Lord of the Rings",
        description: "Complete trilogy book set in hardcover. Includes beautiful illustrations and maps. Like new condition.",
        price: "199.99",
        image_filename: null,
        category_id: 2,
        quantity: 10,
        status: "active",
        created_at: "2025-05-09T10:00:00Z",
        updated_at: null,
        seller_id: "550e8400-e29b-41d4-a716-446655440001",
        seller: {
          id: "550e8400-e29b-41d4-a716-446655440001",
          first_name: "John",
          last_name: "Seller"
        },
        category: {
          id: 2,
          name: "Books"
        }
      },
      "550e8400-e29b-41d4-a716-446655440003": {
        id: "550e8400-e29b-41d4-a716-446655440003",
        title: "Nike Running Shoes",
        description: "Premium running shoes size 42. Very comfortable for long distance running. Used only a few times, excellent condition.",
        price: "299.99",
        image_filename: null,
        category_id: 3,
        quantity: 3,
        status: "active",
        created_at: "2025-05-08T14:30:00Z",
        updated_at: null,
        seller_id: "550e8400-e29b-41d4-a716-446655440004",
        seller: {
          id: "550e8400-e29b-41d4-a716-446655440004",
          first_name: "Alice",
          last_name: "Vendor"
        },
        category: {
          id: 3,
          name: "Clothing"
        }
      }
    };

    const offer = mockOffers[offerId];
    
    if (!offer) {
      throw new Error('Offer not found');
    }
    
    return offer;
  } catch (error) {
    console.error('Error fetching offer details:', error);
    throw error;
  }
};

// Helper function to get category name by ID
const getCategoryNameById = (categoryId) => {
  const categoryMap = {
    1: "Electronics",
    2: "Books",
    3: "Clothing",
    4: "Home",
    5: "Games"
  };
  return categoryMap[categoryId] || "Unknown Category";
};

/**
 * Updates an existing offer
 * @param {string} offerId - ID of the offer to update
 * @param {FormData} formData - Form data containing updated offer details
 * @returns {Promise<Object>} Updated offer data
 */
export const updateOffer = async (offerId, formData) => {
  try {
    const response = await api.put(`/offers/${offerId}`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    });
    return response.data;
  } catch (error) {
    console.error('Error updating offer:', error);
    throw new Error(
      error.response?.data?.message || 'Nie udało się zaktualizować oferty'
    );
  }
};

/**
 * Changes the status of an offer
 * @param {string} offerId - ID of the offer
 * @param {string} status - New status (active, inactive, etc.)
 * @returns {Promise<Object>} Updated offer data
 */
export const changeOfferStatus = async (offerId, status) => {
  try {
    const response = await api.patch(`/offers/${offerId}/status`, { status });
    return response.data;
  } catch (error) {
    console.error('Error changing offer status:', error);
    throw new Error(
      error.response?.data?.message || 'Nie udało się zmienić statusu oferty'
    );
  }
};

File: frontend/src/services/offer_service.py
from datetime import datetime
from decimal import Decimal
from logging import Logger
from typing import Optional
from uuid import UUID

from fastapi import BackgroundTasks, HTTPException, UploadFile, status
from sqlalchemy import case, func
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select

from exceptions.offer_exceptions import (InvalidStatusTransitionException,
                                         NotOfferOwnerException,
                                         OfferAlreadyInactiveException,
                                         OfferAlreadySoldException,
                                         OfferModificationFailedException,
                                         OfferNotFoundException)
from models import CategoryModel, LogModel, OfferModel, UserModel
from schemas import (CategoryDTO, LogEventType, OfferDetailDTO,
                     OfferListResponse, OfferStatus, OfferSummaryDTO,
                     SellerInfoDTO, UserRole)
from services.file_service import FileService
from utils.pagination_utils import build_paginated_response


# Funkcja pomocnicza do logowania zdarzeń
async def log_event(
    db_session: AsyncSession,
    event_type: LogEventType,
    user_id: UUID,
    message: str,
) -> None:
    """
    Helper function to log events in the system

    Args:
        db_session: Database session
        event_type: Type of the event from LogEventType enum
        user_id: ID of the user who triggered the event
        message: Event description message
    """
    log_entry = LogModel(
        event_type=event_type, user_id=user_id, message=message
    )
    db_session.add(log_entry)
    await db_session.commit()


class OfferService:
    def __init__(self, db_session: AsyncSession, logger: Logger):
        self.db_session = db_session
        self.logger = logger
        self.file_service = FileService(logger)

    async def create_offer(
        self,
        seller_id: UUID,
        title: str,
        price: Decimal,
        category_id: int,
        quantity: int = 1,
        description: Optional[str] = None,
        image: Optional[UploadFile] = None,
        background_tasks: Optional[BackgroundTasks] = None,
    ) -> OfferSummaryDTO:
        """
        Creates a new offer in the database.

        Args:
            seller_id: ID of the seller creating the offer
            title: Title of the offer
            price: Price of the product
            category_id: ID of the product category
            quantity: Quantity available (default 1)
            description: Optional product description
            image: Optional product image
            background_tasks: Optional background tasks for async operations

        Returns:
            OfferSummaryDTO: The created offer details

        Raises:
            HTTPException: If validation fails or database operation fails
        """
        try:
            # Validate price
            if price <= 0:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail={
                        "error_code": "INVALID_PRICE",
                        "message": "Price must be greater than 0",
                    },
                )

            # Validate quantity
            if quantity < 0:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail={
                        "error_code": "INVALID_QUANTITY",
                        "message": "Quantity cannot be negative",
                    },
                )

            # Verify category exists
            result = await self.db_session.execute(
                select(CategoryModel).where(CategoryModel.id == category_id)
            )
            category = result.scalars().first()

            if not category:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail={
                        "error_code": "CATEGORY_NOT_FOUND",
                        "message": "Category not found",
                    },
                )

            # Process image if provided
            image_filename = None
            if image:
                image_filename = await self.file_service.save_image(image)

            # Create offer in database
            new_offer = OfferModel(
                seller_id=seller_id,
                category_id=category_id,
                title=title,
                description=description,
                price=price,
                quantity=quantity,
                image_filename=image_filename,
                status=OfferStatus.INACTIVE,  # Default status is inactive
            )

            self.db_session.add(new_offer)

            # Log event
            log_entry = LogModel(
                event_type=LogEventType.OFFER_CREATE,
                user_id=seller_id,
                message=f"User {seller_id} created a new offer: {title}",
            )
            self.db_session.add(log_entry)

            await self.db_session.commit()
            await self.db_session.refresh(new_offer)

            # Convert to DTO and return
            return OfferSummaryDTO(
                id=new_offer.id,
                seller_id=new_offer.seller_id,
                category_id=new_offer.category_id,
                title=new_offer.title,
                price=new_offer.price,
                image_filename=new_offer.image_filename,
                quantity=new_offer.quantity,
                status=new_offer.status,
                created_at=new_offer.created_at,
            )

        except HTTPException:
            await self.db_session.rollback()
            raise
        except Exception as e:
            await self.db_session.rollback()
            self.logger.error(f"Error creating offer: {str(e)}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail={
                    "error_code": "CREATE_FAILED",
                    "message": "Failed to create the offer",
                },
            )

    async def deactivate_offer(
        self, offer_id: UUID, user_id: UUID, user_role: UserRole
    ) -> OfferDetailDTO:
        """
        Deactivate an offer by changing its status from 'active' to 'inactive'

        Args:
            offer_id: UUID of the offer to deactivate
            user_id: UUID of the current user
            user_role: Role of the current user

        Returns:
            OfferDetailDTO with the updated offer details

        Raises:
            HTTPException: Various exceptions based on validation and authorization
        """
        try:
            # Verify user is a seller
            if user_role != UserRole.SELLER:
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail={
                        "error_code": "INSUFFICIENT_PERMISSIONS",
                        "message": "Only sellers can deactivate offers",
                    },
                )

            # Get the offer
            offer = await self.db_session.get(OfferModel, offer_id)

            # Check if offer exists
            if not offer:
                raise OfferNotFoundException(offer_id)

            # Check if user is the owner
            if offer.seller_id != user_id:
                raise NotOfferOwnerException(offer_id)

            # Check if offer is already inactive
            if offer.status == OfferStatus.INACTIVE:
                raise OfferAlreadyInactiveException(offer_id)

            # Check if offer status allows deactivation (must be 'active')
            if offer.status != OfferStatus.ACTIVE:
                raise InvalidStatusTransitionException(
                    current_status=offer.status,
                    target_status=OfferStatus.INACTIVE,
                )

            # Update offer status
            offer.status = OfferStatus.INACTIVE
            offer.updated_at = datetime.now()

            # Save to database
            await self.db_session.commit()
            await self.db_session.refresh(offer)

            # Log the status change event
            await log_event(
                self.db_session,
                LogEventType.OFFER_STATUS_CHANGE,
                user_id,
                f"Offer {offer_id} deactivated",
            )

            # Fetch seller and category for full response
            seller = await self.db_session.get(UserModel, offer.seller_id)
            category = await self.db_session.get(
                CategoryModel, offer.category_id
            )

            # Return detailed DTO
            return self._map_to_offer_detail_dto(offer, seller, category)

        except (
            OfferNotFoundException,
            NotOfferOwnerException,
            InvalidStatusTransitionException,
            OfferAlreadyInactiveException,
        ):
            # Re-raise custom exceptions
            raise
        except HTTPException:
            # Re-raise HTTP exceptions
            raise
        except Exception as e:
            # Log the error
            self.logger.error(f"Error deactivating offer: {str(e)}")
            raise OfferModificationFailedException(
                operation="deactivate", details=str(e)
            )

    def _map_to_offer_detail_dto(self, offer, seller, category):
        """
        Maps database models to OfferDetailDTO

        Args:
            offer: OfferModel instance
            seller: UserModel instance
            category: CategoryModel instance

        Returns:
            OfferDetailDTO with offer details, seller and category information
        """
        seller_info = SellerInfoDTO(
            id=seller.id,
            first_name=seller.first_name,
            last_name=seller.last_name,
        )

        category_info = CategoryDTO(id=category.id, name=category.name)

        return OfferDetailDTO(
            id=offer.id,
            seller_id=offer.seller_id,
            category_id=offer.category_id,
            title=offer.title,
            description=offer.description,
            price=offer.price,
            image_filename=offer.image_filename,
            quantity=offer.quantity,
            status=offer.status,
            created_at=offer.created_at,
            updated_at=offer.updated_at,
            seller=seller_info,
            category=category_info,
        )

    async def mark_offer_as_sold(
        self, offer_id: UUID, user_id: UUID, user_role: UserRole
    ) -> OfferDetailDTO:
        """
        Mark an offer as sold by changing its status to 'sold' and setting quantity to 0

        Args:
            offer_id: UUID of the offer to mark as sold
            user_id: UUID of the current user
            user_role: Role of the current user

        Returns:
            OfferDetailDTO with the updated offer details

        Raises:
            HTTPException: Various exceptions based on validation and authorization
        """
        try:
            # Verify user is a seller
            if user_role != UserRole.SELLER:
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail={
                        "error_code": "INSUFFICIENT_PERMISSIONS",
                        "message": "Only sellers can mark offers as sold",
                    },
                )

            # Get the offer
            offer = await self.db_session.get(OfferModel, offer_id)

            # Check if offer exists
            if not offer:
                raise OfferNotFoundException(offer_id)

            # Check if user is the owner
            if offer.seller_id != user_id:
                raise NotOfferOwnerException(offer_id)

            # Check if offer is already sold
            if offer.status == OfferStatus.SOLD:
                raise OfferAlreadySoldException(offer_id)

            # Check if offer status allows marking as sold (cannot be archived or deleted)
            if offer.status in [OfferStatus.ARCHIVED, OfferStatus.DELETED]:
                raise InvalidStatusTransitionException(
                    current_status=offer.status, target_status=OfferStatus.SOLD
                )

            # Update offer status and quantity
            offer.status = OfferStatus.SOLD
            offer.quantity = 0
            offer.updated_at = datetime.now()

            # Save to database
            await self.db_session.commit()
            await self.db_session.refresh(offer)

            # Log the status change event
            await log_event(
                self.db_session,
                LogEventType.OFFER_STATUS_CHANGE,
                user_id,
                f"Offer {offer_id} marked as sold",
            )

            # Fetch seller and category for full response
            seller = await self.db_session.get(UserModel, offer.seller_id)
            category = await self.db_session.get(
                CategoryModel, offer.category_id
            )

            # Return detailed DTO
            return self._map_to_offer_detail_dto(offer, seller, category)

        except (
            OfferNotFoundException,
            NotOfferOwnerException,
            InvalidStatusTransitionException,
            OfferAlreadySoldException,
        ):
            # Re-raise custom exceptions
            raise
        except HTTPException:
            # Re-raise HTTP exceptions
            raise
        except Exception as e:
            # Log the error
            self.logger.error(f"Error marking offer as sold: {str(e)}")
            raise OfferModificationFailedException(
                operation="mark_sold", details=str(e)
            )

    async def list_all_offers(
        self,
        search: Optional[str] = None,
        category_id: Optional[int] = None,
        seller_id: Optional[UUID] = None,
        status: Optional[OfferStatus] = None,
        sort: str = "created_at_desc",
        page: int = 1,
        limit: int = 100,
    ) -> OfferListResponse:
        """
        Retrieve a paginated list of all offers for admin, with filtering, sorting, and pagination.
        """
        try:
            # Calculate offset
            offset = (page - 1) * limit

            # Base query
            base_query = select(OfferModel)
            filters = []
            if search:
                term = f"%{search}%"
                filters.append(
                    or_(
                        OfferModel.title.ilike(term),
                        OfferModel.description.ilike(term),
                    )
                )
            if category_id is not None:
                filters.append(OfferModel.category_id == category_id)
            if seller_id is not None:
                filters.append(OfferModel.seller_id == seller_id)
            if status is not None:
                filters.append(OfferModel.status == status)

            # Apply filters
            query = base_query.where(*filters) if filters else base_query

            # Sorting
            if sort == "price_asc":
                query = query.order_by(OfferModel.price.asc())
            elif sort == "price_desc":
                query = query.order_by(OfferModel.price.desc())
            elif sort == "relevance" and search:
                # Simple relevance: exact match first, prefix match next, then others
                # Using term for ILIKE conditions
                # term is defined above as f"%{search}%"
                query = query.order_by(
                    case(
                        [
                            (OfferModel.title.ilike(term), 1),
                            (OfferModel.title.ilike(f"{search}%"), 2),
                        ],
                        else_=3,
                    )
                )
            else:
                # default sort by creation date descending
                query = query.order_by(OfferModel.created_at.desc())

            # Count total
            count_q = select(func.count()).select_from(query.subquery())
            total = (await self.db_session.execute(count_q)).scalar() or 0

            # Fetch paginated data
            result = await self.db_session.execute(
                query.offset(offset).limit(limit)
            )
            offers = result.scalars().all()

            # Map to DTOs
            items = [
                OfferSummaryDTO(
                    id=o.id,
                    seller_id=o.seller_id,
                    category_id=o.category_id,
                    title=o.title,
                    price=o.price,
                    image_filename=o.image_filename,
                    quantity=o.quantity,
                    status=o.status,
                    created_at=o.created_at,
                )
                for o in offers
            ]

            # Build and return paginated response
            paginated = build_paginated_response(items, total, page, limit)
            # Convert to OfferListResponse
            return OfferListResponse(**paginated.dict())
        except Exception as e:
            self.logger.error(f"Error fetching all offers: {str(e)}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail={
                    "error_code": "FETCH_FAILED",
                    "message": "Failed to fetch offers",
                },
            )

    async def moderate_offer(self, offer_id: UUID) -> OfferDetailDTO:
        """
        Moderates an offer by changing its status to 'moderated'.
        """
        # Retrieve the offer
        offer = await self.db_session.get(OfferModel, offer_id)
        if not offer:
            raise OfferNotFoundException(offer_id)
        if offer.status == OfferStatus.MODERATED:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail={
                    "error_code": "ALREADY_MODERATED",
                    "message": "Offer is already moderated",
                },
            )
        # Update status and timestamp
        offer.status = OfferStatus.MODERATED
        offer.updated_at = datetime.now()
        try:
            await self.db_session.commit()
            await self.db_session.refresh(offer)
        except HTTPException:
            await self.db_session.rollback()
            raise
        except Exception as e:
            await self.db_session.rollback()
            self.logger.error(f"Error moderating offer {offer_id}: {str(e)}")
            raise OfferModificationFailedException(
                operation="moderation", details=str(e)
            )
        # Fetch related entities for full DTO
        seller = await self.db_session.get(UserModel, offer.seller_id)
        category = await self.db_session.get(CategoryModel, offer.category_id)
        return self._map_to_offer_detail_dto(offer, seller, category)

    async def unmoderate_offer(self, offer_id: UUID) -> OfferDetailDTO:
        """
        Unmoderates an offer by changing its status from 'moderated' to 'inactive'.
        """
        # Retrieve the offer
        offer = await self.db_session.get(OfferModel, offer_id)
        if not offer:
            raise OfferNotFoundException(offer_id)
        # Ensure the offer is currently moderated
        if offer.status != OfferStatus.MODERATED:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail={
                    "error_code": "NOT_MODERATED",
                    "message": "Offer is not moderated",
                },
            )
        # Update status and timestamp
        offer.status = OfferStatus.INACTIVE
        offer.updated_at = datetime.now()
        try:
            await self.db_session.commit()
            await self.db_session.refresh(offer)
        except HTTPException:
            await self.db_session.rollback()
            raise
        except Exception as e:
            await self.db_session.rollback()
            self.logger.error(f"Error unmoderating offer {offer_id}: {str(e)}")
            raise OfferModificationFailedException(
                operation="unmoderation", details=str(e)
            )
        # Fetch related entities for full DTO
        seller = await self.db_session.get(UserModel, offer.seller_id)
        category = await self.db_session.get(CategoryModel, offer.category_id)
        return self._map_to_offer_detail_dto(offer, seller, category)
